src/lib/uTop_main.ml

```diff
@@ -79,6 +79,39 @@ let index_of_offset src ofs =
 
 let convert_locs str locs = List.map (fun (a, b) -> (index_of_offset str a, index_of_offset str b)) locs
 
+let convert_loc str (a, b) = (index_of_offset str a, index_of_offset str b)
+
+let get_line src line =
+  let rec aux line' ofs skipped =
+    if ofs >= String.length src then
+      ("", 0)
+    else if line' = line then
+      (String.sub src ofs (String.length src - ofs), skipped)
```

Comment:
How come you aren't using `Zed_utf8.sub` and `Zed_utf8.length` here?

Reply:
It is not clear to me how that can be useful. This function finds the when a new line starts and keeps track of how many characters are before it. I could not pass around skipped and use Zed_utf8.length instead to calculate it. However, that would mean redo a traversal of everything because UTF8.

I agree this code is a bit tricky. IIRC my first implementation was to split the str in lines so you can address easily by line. It was much more memory inefficient but, after all, utop works on small chunks of code. I will submit this piece of code in the new PR and we can move the discussion there. I think the best way to make this more readable is to use the trivial approach to split into lines and have them directly addressable.